packages/dds/tree/docs/wip/cell-model.md

```diff
@@ -0,0 +1,210 @@
+Dynamically-sized fields typically start out empty of any cell.
+
+The low-level edits that such fields might support can be decomposed as follows:
+
+- Insert: allocate and fill a cell
+
+- Delete/Remove: clear a cell
+
+- Revive: fill a (cleared) cell with the contents it contained before
+
+- Replace: fill a (filled) cell
+
+- Move content from A to B:
+  
+  - A: clear a (filled) cell and add a forwarding annotation to it
+  
+  - B: allocate and fill a cell with the content being moved
+
+- Return content to A from B:
+  
+  - A: remove the forwarding annotation from a cell and fill it with the returned content
+  
+  - B: clear the (filled) cell
+
+### Edits on Fixed-Sized Fields
+
+Fixed-sized fields can be used to model required (i.e., unary) fields, optional fields, fixed-sized arrays (not to be confused with JavaScript arrays which behave like dynamically-sized lists), and tuples.
+Fixed-sized fields are populated with their cells (and possibly content) from the time they are created.
+
+The low-level edits that such fields might support can be decomposed as follows:
+
+- Upsert: fill a (potentially empty) cell
+
+- Populate if empty: fill a cell if the cell is empty
+
+- Load-linked store: fill a cell if the cell has not been concurrently filled or cleared
+
+- Clear: clear a cell
+
+Note that no further cell allocations are made after the field is created.
+
+## Implications For SharedTree
+
+### Data Model
+
+The SharedTree data model should differentiate between fixed-sized and dynamically-sized fields.
+Users of SharedTree, when writing a schema, implicitly choose whether to use fixed-size or dynamically-sized field.
+This is not a new choice (they're still choosing from the same set of options) but their choice is now reflected at the data model level.
+
+### Editing API
+
+New editing operations need to be exposed for fixed-sized fields.
+
+The editing operations offered by SharedTree, even for unschematized data, need not be the ones prescribed by the model.
+Instead, we see value in offering two sets of higher-level operations: one for fixed-sized fields and one for dynamically-sized fields.
+This allows the editing API to be more specialized for each kind of field, which in turn allows the API to be more expressive and more familiar.
+
+### Merge Semantics
+
+The merge resolution logic can now exclude the possibility of having to merge operations from both sets within a single field (outside of schema migration scenarios).
+
+Since the merge resolution logic is specialized to each field kind, merge resolution logic becomes extensible: we can introduce new kind of fields with associated merge logic.
+This allows us to support new kinds of fields in the future, but more importantly it means that we can migrate from a less desirable set of merge semantics to a more desirable one.
+This dramatically reduces the negative impact of starting out is suboptimal the merge semantics.
```

Comment:
Nit: typo? I didn't follow this.

Reply:
Typo indeed. Let me know if the sentence is clear now.

Reply:
Yup! That makes sense!